ssl/s3_lib.c

```diff
@@ -4145,9 +4148,52 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     }
 #endif
 
-    if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
+    /* SUITE-B takes precedence over server preference and ChaCha priortiy */
+    if (tls1_suiteb(s)) {
+        prio = srvr;
+        allow = clnt;
+    } else if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha20 is at the top of the client preference list,
+           and there are ChaCha20 ciphers in the server list, then
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, check server... */
+                int num = sk_SSL_CIPHER_num(srvr);
+                int found = 0;
+                for (i = 0; i < num; i++) {
+                    c = sk_SSL_CIPHER_value(srvr, i);
+                    if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                        found = 1;
+                        break;
+                    }
+                }
+                if (found) {
+                    new_chacha = sk_SSL_CIPHER_new_null();
+                    /* if reserve fails, then there's likely a memory issue */
+                    if (new_chacha != NULL && sk_SSL_CIPHER_reserve(new_chacha, num)) {
+                        /* Put ChaCha at the top */
+                        for (i = 0; i < num; i++) {
+                            c = sk_SSL_CIPHER_value(srvr, i);
+                            if (c->algorithm_enc == SSL_CHACHA20POLY1305)
+                                sk_SSL_CIPHER_push(new_chacha, c);
+                        }
```

Comment:
??? You already have `c` to push, you don't need this loop at all, just sk_SSL_CIPHER_push(new_chacha, c)...

Reply:
Yes, it's needed, there can be more than one ChaCha20 cipher in the server list, and they all need to be put at the start. The first loop at 4168~4174 attempts to avoid allocating the new stack, if possible.

Reply:
I suppose I could shorten the loop by pushing the found cipher first; then iterating through the rest.

Reply:
Ah! You mean that there are other fields in client's head of list and goal is to prioritize *all* Chachas. One could omit this loop if the match was exact...

Reply:
We're still doing server priority; there's a decent chance (>=50% given the number of ChaCha20 ciphers) that they won't match, but another ChaCha20 cipher in the list will.

Reply:
Question is if prioritizing *all* Chachas is actually what client wants. Would maybe starting to push ciphers from client list till it's not Chacha, and then append server list omitting all Chachas more adequate?

Reply:
No, this would be problematic, right? Because in this case you wouldn't be able to interchange prio and allow...

Reply:
We still want server priority, not client priority; so we want to only include those ChaCha ciphers in the server list, in the order that the server has them.

Reply:
All right, rename new_chacha to prio_chacha, *optionally* start loop from i+1 instead of 0 (you'll have to push c first), and add comment explaining that we pull up all Chacha ciphers.

Reply:
Done